sched/scheduler/stateful_scheduler.go

```diff
@@ -578,26 +576,33 @@ addLoop:
 		select {
 		case newJobMsg := <-s.addJobCh:
 			receivedJob = true
-
+			lf := log.Fields{
+				"jobID":     newJobMsg.job.Id,
+				"requestor": newJobMsg.job.Def.Requestor,
+				"jobType":   newJobMsg.job.Def.JobType,
+				"tag":       newJobMsg.job.Def.Tag,
+				"basis":     newJobMsg.job.Def.Basis,
+				"priority":  newJobMsg.job.Def.Priority,
+				"numTasks":  len(newJobMsg.job.Def.Tasks),
+			}
 			if _, ok := s.requestorMap[newJobMsg.job.Def.Requestor]; ok {
-				// If we have an existing job with this requestor/tag combination, make sure we use its priority level.
-				// Not an error since we can consider priority to be a suggestion which we'll handle contextually.
 				for _, js := range s.requestorMap[newJobMsg.job.Def.Requestor] {
+					// Kill any prior jobs that share the same requestor and non-empty basis string,
+					// but tag should be different so we don't kill the original jobs when doing retries.
+					if newJobMsg.job.Def.Basis != "" &&
+						js.Job.Def.Basis == newJobMsg.job.Def.Basis &&
+						js.Job.Def.Tag != newJobMsg.job.Def.Tag {
+						log.WithFields(lf).Infof("Matching basis, killing jobID=%s", js.Job.Id)
+						s.KillJob(js.Job.Id)
+					}
+
+					// If we have an existing job with this requestor/tag combination, make sure we use its priority level.
+					// Not an error since we can consider priority to be a suggestion which we'll handle contextually.
 					if js.Job.Def.Tag == newJobMsg.job.Def.Tag &&
 						js.Job.Def.Basis != newJobMsg.job.Def.Basis &&
 						js.Job.Def.Priority != newJobMsg.job.Def.Priority {
-						m := newJobMsg
-						log.WithFields(
-							log.Fields{
-								"requestor": m.job.Def.Requestor,
-								"jobType":   m.job.Def.JobType,
-								"tag":       m.job.Def.Tag,
-								"basis":     m.job.Def.Basis,
-								"priority":  m.job.Def.Priority,
-								"numTasks":  len(m.job.Def.Tasks),
-							}).Info("Overriding job priority to match previous requestor/tag priority")
+						log.WithFields(lf).Info("Overriding job priority to match previous requestor/tag priority")
```

Comment:
was the 'break' a bug before?

Reply:
Since new retry jobs have to match previous jobs, would could exit early before, but now we want to unconditionally iterate across all jobs with the same requestor.